include/ofi_uffd.h

```diff
@@ -0,0 +1,209 @@
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _OFI_UFFD_H_
+#define _OFI_UFFD_H_
+
+#if HAVE_LINUX_USERFAULTFD_H
+
+#include <config.h>
+#include <stdlib.h>
+#include <ofi_util.h>
+#include <errno.h>
+#include <sys/syscall.h>
+#include <sys/ioctl.h>
+#include <linux/userfaultfd.h>
+
+#define MB(x)((size_t)(x) << 20)
+#define GB(x)((size_t)(x) << 30)
+
+/* See TODO in function ofi_uffd_register */
+static long page_sizes[] = {4096, MB(2), GB(1)};
+
+static long get_page_size(void)
+{
+	long ret;
+
+	/* sysconf can return -1 and not change errno */
+	errno = 0;
+	ret = sysconf(_SC_PAGESIZE);
+
+	if (ret <= 0) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+			"sysconf/pagesize: ret %ld, err %s\n",
+			ret, strerror(errno));
+		if (errno)
+			return -errno;
+		else
+			return -FI_EOTHER;
+	}
+
+	return ret;
+}
+
+static inline void set_uffdio_range(struct uffdio_range *range,
+				    uint64_t start, uint64_t len,
+				    long page_size)
+{
+	range->start = (start / page_size) * page_size;
+	range->len = ofi_div_ceil(len, page_size) * page_size;
+}
+
+/*
+ * TODO: ioctl register needs to be passed page aligned boundaries.
+ * Need to find a more sane way of detecting a region's page size.
+ */
+static inline int ofi_uffd_register(int uffd, uint64_t start, uint64_t len)
+{
+	struct uffdio_register uffdio_register;
+	long page_size;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(page_sizes); i++) {
+		page_size = page_sizes[i];
+		set_uffdio_range(&uffdio_register.range, start, len, page_size);
+		uffdio_register.mode = UFFDIO_REGISTER_MODE_MISSING;
+
+		if (ioctl(uffd, UFFDIO_REGISTER, &uffdio_register) != -1)
+			return 0;
+
+		if (errno != EINVAL) {
+			FI_WARN(&core_prov, FI_LOG_MR, "ioctl/uffdio_register: %s\n",
+				strerror(errno));
+			break;
+		}
+	}
+
+	return -errno;
+}
+
+static inline int ofi_uffd_unregister(int uffd, uint64_t start, uint64_t len)
+{
+	struct uffdio_range uffdio_range;
+	long page_size;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(page_sizes); i++) {
+		page_size = page_sizes[i];
+		set_uffdio_range(&uffdio_range, start, len, page_size);
+		if (ioctl(uffd, UFFDIO_UNREGISTER, &uffdio_range) != -1)
```

Comment:
I'm not overly familiar with the userfaultd stuff; forgive me if this is a naieve question.

Is it possible that you register region X, but then region X gets unmapped/removed from the process, but then before `ofi_uffd_unregister()` can get called to unregister region X, region X gets mapped again to some other physical address?

I.e., is there atomicity in the notifications of unmapping and the unregistration in this function?

Reply:
This is the general operation:
"When a page fault occurs for one of the regions registered to the userfaultfd object, the faulting thread is put to sleep and an event is generated that can be read via the userfaultfd file descriptor...  the fault-handling thread can trigger a wake-up for the sleeping thread."

So...you need a separate thread monitoring the fd.  Whenever a thread maps or unmaps memory, that thread will block in the kernel until the monitoring thread handles the event and wakes it back up.

Reply:
Probably also need to make 100% sure that the thread doesn't use any registered memory that may become unmapped (because then it will go to sleep).  This may be an obvious "duh", but I figured it was worth being explicit about... 😄